controllers/virtualmachinereplicaset/virtualmachinereplicaset_controller.go

```diff
@@ -202,38 +223,441 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Re
+
+		if err := helper.Patch(ctx, vm); err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to apply %s label to VirtualMachine %q",
+				vmopv1.VirtualMachineReplicaSetNameLabel, vm.Name)
+		}
+
+		// TODO: Propagate the Deployment label from Deployment to replica set to VirtualMachine if
+		// it is set on the replica set.
+	}
+
+	syncErr := r.syncReplicas(ctx, ctx.ReplicaSet, filteredVMs)
+
+	// Update the status of the VirtualMachineReplicaSet since syncing might
+	// have resulted in replicas being added or removed.
+	r.updateStatus(ctx, ctx.ReplicaSet, filteredVMs)
+
+	if syncErr != nil {
+		return ctrl.Result{}, errors.Wrapf(syncErr, "failed to sync VirtualMachineReplicaSet replicas")
+	}
+
+	var replicas int32
+	if ctx.ReplicaSet.Spec.Replicas != nil {
+		replicas = *ctx.ReplicaSet.Spec.Replicas
+	}
+
+	// Queue faster reconciles until all replicas are available.
+	if ctx.ReplicaSet.Status.ReadyReplicas != replicas {
+		return ctrl.Result{RequeueAfter: 15 * time.Second}, nil
+	}
+
+	return ctrl.Result{}, nil
 }
 
-func (r *Reconciler) ReconcileDelete(ctx *pkgctx.VirtualMachineReplicaSetContext) (reterr error) {
-	if controllerutil.ContainsFinalizer(ctx.ReplicaSet, finalizerName) {
-		defer func() {
-			r.Recorder.EmitEvent(ctx.ReplicaSet, "Delete", reterr, false)
-		}()
+// MustEqualValue returns true if the replica set name equals either the label
+// value or its hashed value.
+func MustEqualValue(rsName, labelValue string) bool {
+	return labelValue == util.MustFormatValue(rsName)
+}
 
-		// Handle deletion of replicas
+// getNewVirtualMachine creates a new VirtualMachine object. The name of the newly created resource is going
+// to be created by the API server, we set the generateName field.
+func (r *Reconciler) getNewVirtualMachine(rs *vmopv1.VirtualMachineReplicaSet) *vmopv1.VirtualMachine {
+	gv := vmopv1.SchemeGroupVersion
+	vm := &vmopv1.VirtualMachine{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: fmt.Sprintf("%s-", rs.Name),
+			// Note: by setting the ownerRef on creation we signal to the
+			// VirtualMachine controller that this is not a stand-alone VM.
+			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(rs, replicaSetKind)},
+			Namespace:       rs.Namespace,
+			Labels: map[string]string{
+				vmopv1.VirtualMachineReplicaSetNameLabel: util.MustFormatValue(rs.Name),
+			},
+			Annotations: rs.Spec.Template.Annotations,
+		},
+		TypeMeta: metav1.TypeMeta{
+			Kind:       gv.WithKind("VirtualMachine").Kind,
+			APIVersion: gv.String(),
+		},
+		Spec: rs.Spec.Template.Spec,
+	}
 
-		controllerutil.RemoveFinalizer(ctx.ReplicaSet, finalizerName)
+	// Copy over the labels specified in replicaset's template to the
+	// VirtualMachine. Note that a direct assignment here would share the labels
+	// and consequently, any changes to the VM label will also impact the
+	// replicaset's template labels.
+	for k, v := range rs.Spec.Template.Labels {
```

Comment:
`maps.Copy()`?

Do we make sure that `rs.Spec.Template.Labels` does not contain `VirtualMachineReplicaSetNameLabel` ?

Reply:
Thanks for this.

> Do we make sure that rs.Spec.Template.Labels does not contain VirtualMachineReplicaSetNameLabel

Hmm.  I think that should be okay?  If the template labels contain the replicaset name as a label, that will also be present as a selector on the repilcaset as well -- the webhook ensures that the `rs.spec.Selector.MatchLabels == rs.Spec.Template.Labels`.  I also don't see a validation like that in upstream k8s replicaset controller, or the machineset controller in cluster-api.

Reply:
I am going to resolve this for now.  Totally fine with revisiting this if I missed something.